utils.py

```diff
@@ -78,3 +97,40 @@ def calculate_total_karma(word: str, last_words: deque[str]) -> float:
     decay: float = calculate_decay(n)
     base_karma: float = calculate_base_karma(word)
     return decay * base_karma if base_karma > 0 else base_karma
+
+# =================================================================================================================
+
+
+@contextlib.asynccontextmanager
+async def db_connection(bot: WordChainBot, locked: bool = True) -> AsyncIterator[AsyncConnection]:
```

Comment:
What's the motivation to move it away from the bot?
IDEs correctly complain here, that you shouldn't access private or protected members outside of it's own or a derived class.
It also forces you to do the `if TYPE_CHECKING:` workaround above to avoid cyclic imports.

Reply:
True, true. This is why a second pair of eyes helps. I had initially thought of getting the db connection differently instead of using the bot's SQL engine and lock. But in the end, I ended up having access to the bot's instance everywhere, so it makes sense now to send it back to the main bot's class.